### DISCUSSION TITLE
function structure for unit tests

### DISCUSSION THREAD

**Discussion by sebsch1337** [neuefische/questions#298] [Web] [answered]
Hello everyone,

at the moment, my function logic (toggleModal, onDeleteDestination, calculateTotalDuration, ...) is stored in its regarding page. (view example below)

Now i try to implement unit tests for those functions. But since they are nested functions, it is not possible to export them for testing.

What is the best way to implement unit tests in this case? Should i move the functions into a different scope, like above `export default function Destinations...` and handle the in-scope-states with attributes? (`const calculateTotalDuration(destinations)`)

Or is it even better to write an own JS-file for these functions and import them into the regarding Destinations component and unit test?

Thanks in advance!

```
export default function Destinations() {
  const router = useRouter();
  const { id } = router.query;
  const [trips, setTrips] = useLocalStorage("trips", dummyTrips);
  const [destinations, setDestinations] = useLocalStorage("destinations", dummyDestinations);
  const [modal, setModal] = useState({ visible: false, name: "", id: "" });

  const calculateTotalDuration = () => {
    const minDate = Math.min(
      ...destinations.filter((filtered) => filtered.tripId === id).map((destination) => destination.startDate)
    );
    const maxDate = Math.max(
      ...destinations.filter((filtered) => filtered.tripId === id).map((destination) => destination.endDate)
    );
    const dayDifference = (maxDate - minDate) / 86400 + 1;
    return dayDifference > 0 ? dayDifference : 0;
  };
```

**Comment by ThomasF85** [comment, marked as answer]
In this case you can put the function outside of the component and give it two parameters, like this:
```
export const calculateTotalDuration = (destinations, id) => {
  ...
};
```
If you need to calculate the total duration again elsewhere you could put it into a separate file with utility functions.

**Reply by sebsch1337** [reply]
Would you generally recommend this approach (working with parameters instead of accessing states directly from the higher scope)?

**Reply by ThomasF85** [reply]
In most of your usual cases (writing functions that manipulate a state) it is totally fine having a function accessing state directly.
